# Capstone Project: IBM-Coursera Professional Data Science Specialization

#####      Introduction

This project is designed to answer the question: Is there a Toronto neighbourhood that has a sufficient number of the qualities I love about my own Philadelphia, PA neighborhood - as well as a promising business environment where my firm (BECG, LLC) could be positioned for growth?

This is my project notebook. I have also submitted a report on Github, and have posted this information to my blog.

## 1. Import, install and load.

In [7]:
!pip install geopy

In [8]:
import numpy as np
import requests
from bs4 import BeautifulSoup
import urllib.request
import pandas as pd
from pandas.io.json import json_normalize

!conda install -c conda-forge geopy --yes

import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be UPDATED:

    ca-certificates: 2019.5.15-0       --> 2019.6.16-hecc5488_0 conda-forge
    certifi:         2019.6.16-py36_0  --> 2019.6.16-py36_1     conda-forge

The following packages will be DOWNGRADED:

    openssl:         1.1.1c-h7b6447c_1 --> 1.1.1c-h516909a_0    conda-forge

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requested packages already installed.



In [9]:
!conda update folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be UPDATED:

    openssl:         1.1.1c-h516909a_0    conda-forge --> 1.1.1c-h7b6447c_1

The following packages will be DOWNGRADED:

    ca-certificates: 2019.6.16-hecc5488_0 conda-forge --> 2019.5.15-0      
    certifi:         2019.6.16-py36_1     conda-forge --> 2019.6.16-py36_0 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done


## 2.  Find data from the Web page containing Philadelphia Zip codes, scrape into Jupyter notebook

In [14]:
#Gets the url and scrapes the html 
url = 'http://ciclt.net/sn/clt/capitolimpact/gw_ziplist.aspx?ClientCode=capitolimpact&State=pa&StName=Pennsylvania&StFIPS=42&FIPS=42101'
req = urllib.request.urlopen(url)


soup = BeautifulSoup(req)

In [15]:
#Finds the table to scrape
table = soup.find('table')

#Provides the empty arrays for the html tags that are being grabbed and assigned to the headings
P = []
C = []

for row in table.find_all('tr'):
    cells = row.find_all('td')
    if len(cells) == 3:
        P.append(cells[0].find(text=True))
        C.append(cells[1].find(text=True))

## 3. Merge Zip code numbers into City data

In [16]:
#Creates the dataframe and places the data in its respective columns
df_phil = pd.DataFrame(P, columns=['PostalCode'])
df_phil['City'] = C
df_phil.head(47)

PostalCode              City
0       19019      Philadelphia
1       19101      Philadelphia
2       19102     Mid City East
3       19102  Middle City East
4       19102      Philadelphia
5       19103     Mid City West
6       19103  Middle City West
7       19103      Philadelphia
8       19104      Philadelphia
9       19105      Philadelphia
10      19106      Philadelphia
11      19107      Philadelphia
12      19108      Philadelphia
13      19109      Philadelphia
14      19110      Philadelphia
15      19111      Philadelphia
16      19112      Philadelphia
17      19114      Philadelphia
18      19115      Philadelphia
19      19116      Philadelphia
20      19118      Philadelphia
21      19119      Philadelphia
22      19120      Philadelphia
23      19121      Philadelphia
24      19122      Philadelphia
25      19123      Philadelphia
26      19124      Philadelphia
27      19125      Philadelphia
28      19126      Philadelphia
29      19127          Manayunk
30      19127      Philadelphia
31      19128      Philadelphia
32      19129      Philadelphia
33      19130      Philadelphia
34      19131      Philadelphia
35      19132      Philadelphia
36      19133      Philadelphia
37      19134      Philadelphia
38      19135      Philadelphia
39      19136      Philadelphia
40      19137      Philadelphia
41      19138      Philadelphia
42      19139      Philadelphia
43      19140      Philadelphia
44      19141      Philadelphia
45      19142      Philadelphia
46      19143      Philadelphia

In [17]:
#A specialized function that joins the neighborhoods with the same postalcode
foo = lambda a: ','.join(a) 
df_phil = df_phil.groupby(['PostalCode']).agg({
                                'City': foo}).reset_index()

In [18]:
df_phil.head()

PostalCode                                         City
0      19019                                 Philadelphia
1      19101                                 Philadelphia
2      19102  Mid City East,Middle City East,Philadelphia
3      19103  Mid City West,Middle City West,Philadelphia
4      19104                                 Philadelphia

## 4. Read the csv file containing latitude and longitude for each Zip code.

In [19]:
df_lonlat = pd.read_csv('https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/download/?format=csv&refine.state=PA&q=philadelphia&timezone=America/New_York&use_labels_for_header=true', delimiter=';')
df_lonlat.head(47)


Zip              City State   Latitude  Longitude  Timezone  \
0   19173      Philadelphia    PA  40.001811 -75.117870        -5   
1   19134      Philadelphia    PA  39.991712 -75.111160        -5   
2   19115      Philadelphia    PA  40.092610 -75.041180        -5   
3   19192      Philadelphia    PA  39.951112 -75.167622        -5   
4   19155      Philadelphia    PA  40.001811 -75.117870        -5   
5   19183      Philadelphia    PA  40.001811 -75.117870        -5   
6   19185      Philadelphia    PA  40.001811 -75.117870        -5   
7   19136      Philadelphia    PA  40.041111 -75.026440        -5   
8   19109      Philadelphia    PA  39.949612 -75.163722        -5   
9   19175      Philadelphia    PA  39.990562 -75.129570        -5   
10  17959  New Philadelphia    PA  40.731739 -76.127800        -5   
11  19104      Philadelphia    PA  39.961612 -75.199570        -5   
12  19146      Philadelphia    PA  39.938512 -75.180670        -5   
13  19105      Philadelphia    PA  40.001811 -75.117870        -5   
14  19101      Philadelphia    PA  40.001811 -75.117870        -5   
15  19131      Philadelphia    PA  39.981112 -75.224860        -5   
16  19177      Philadelphia    PA  40.001811 -75.117870        -5   
17  19107      Philadelphia    PA  39.952112 -75.158530        -5   
18  19147      Philadelphia    PA  39.936562 -75.154090        -5   
19  19171      Philadelphia    PA  40.001811 -75.117870        -5   
20  19154      Philadelphia    PA  40.091460 -74.977190        -5   
21  19196      Philadelphia    PA  40.001811 -75.117870        -5   
22  19127      Philadelphia    PA  40.026626 -75.223110        -5   
23  19124      Philadelphia    PA  40.017362 -75.087690        -5   
24  19191      Philadelphia    PA  40.001811 -75.117870        -5   
25  19132      Philadelphia    PA  39.995412 -75.169770        -5   
26  19148      Philadelphia    PA  39.919812 -75.158030        -5   
27  19182      Philadelphia    PA  40.001811 -75.117870        -5   
28  19111      Philadelphia    PA  40.057661 -75.080180        -5   
29  19119      Philadelphia    PA  40.053511 -75.188580        -5   
30  19102      Philadelphia    PA  39.952962 -75.165580        -5   
31  19112      Philadelphia    PA  39.895677 -75.190440        -5   
32  19184      Philadelphia    PA  40.001811 -75.117870        -5   
33  19187      Philadelphia    PA  40.001811 -75.117870        -5   
34  19152      Philadelphia    PA  40.059611 -75.048370        -5   
35  19108      Philadelphia    PA  39.959662 -75.160500        -5   
36  19143      Philadelphia    PA  39.944162 -75.227180        -5   
37  19144      Philadelphia    PA  40.034111 -75.172030        -5   
38  19128      Philadelphia    PA  40.038944 -75.221040        -5   
39  19103      Philadelphia    PA  39.952162 -75.174060        -5   
40  19141      Philadelphia    PA  40.035778 -75.144700        -5   
41  19122      Philadelphia    PA  39.977662 -75.143360        -5   
42  19194      Philadelphia    PA  40.001811 -75.117870        -5   
43  19114      Philadelphia    PA  40.064257 -75.001550        -5   
44  19150      Philadelphia    PA  40.072260 -75.171060        -5   
45  19116      Philadelphia    PA  40.117413 -75.015400        -5   
46  19138      Philadelphia    PA  40.055861 -75.156540        -5   

    Daylight savings time flag              geopoint  
0                            1   40.001811,-75.11787  
1                            1   39.991712,-75.11116  
2                            1    40.09261,-75.04118  
3                            1  39.951112,-75.167622  
4                            1   40.001811,-75.11787  
5                            1   40.001811,-75.11787  
6                            1   40.001811,-75.11787  
7                            1   40.041111,-75.02644  
8                            1  39.949612,-75.163722  
9                            1   39.990562,-75.12957  
10                           1    40.731739,-76.1278  
11                           1   

## 4A. Drop unnecessary columns

In [20]:
df_lonlat.drop(['City', 'State', 'Timezone', 'Daylight savings time flag', 'geopoint'], axis=1, inplace=True)
df_lonlat.head()

Zip   Latitude  Longitude
0  19173  40.001811 -75.117870
1  19134  39.991712 -75.111160
2  19115  40.092610 -75.041180
3  19192  39.951112 -75.167622
4  19155  40.001811 -75.117870

In [21]:
df_lonlat.rename(columns={'Zip':'PostalCode'}, inplace=True)

In [22]:
df_phil.PostalCode = df_phil.PostalCode.astype(int)

## 5. Merge the Zip Code dataframe and the latitude / longitude dataframe.

In [23]:
df_phil = pd.merge(df_phil, df_lonlat, on='PostalCode', how='outer')

## 6. Generate map showing Zip codes.

In [24]:
# create map of Philadelphia using latitude and longitude values
map_phil = folium.Map(location=[39.9524152, -75.1635755], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_phil['Latitude'], df_phil['Longitude'], df_phil['City']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_phil)  
    
map_phil

If this map does not render properly on Github, you can see it by clicking on [this link](https://github.com/saracw/Coursera---IBM-Capstone-1/blob/master/PhillyClusters1.png). 

### 6A. Scrape data from Web page listing Philadelphia Zip codes.

In [25]:
#Gets the url and scrapes the html 
url1 = 'https://www.zipdatamaps.com/zipcodes-philadelphia-pa'
req1 = urllib.request.urlopen(url1)

soup1 = BeautifulSoup(req1)

### 6B. Load data into dataframe.

In [27]:
table1 = soup1.find('table', class_='table')

Post = []
Pop = []

for row in table1.find_all('tr'):
    cells = row.find_all('td')
    if len(cells) == 8:
        Post.append(cells[0].find(text=True))
        Pop.append(cells[5].find(text=True))

In [28]:
df_pop = pd.DataFrame(Post, columns=['PostalCode'])
df_pop['Population'] = Pop

In [29]:
df_pop.PostalCode = df_pop.PostalCode.astype(int)

### 6C. Merge the Zip code, latitude, and longitude dataframes with population dataframe.

In [30]:
df_merged = pd.merge(df_phil, df_pop, on='PostalCode', how='outer')

In [31]:
%%debug
cols = [2, 3, 4, 5, ]
df_merged.drop(df_merged.columns[cols], axis=0, inplace=True)

NOTE: Enter 'c' at the ipdb>  prompt to continue execution.
> <string>(2)<module>()

ipdb> c
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/indexes/base.py in __getitem__(self, key)
   3966 
   3967         key = com.values_from_object(key)
-> 3968         result = getitem(key)
   3969         if not is_scalar(result):
   3970             return promote(result)

IndexError: index 5 is out of bounds for axis 1 with size 5


In [32]:
df_merged = df_merged.dropna()

In [33]:
df_merged.head(47)

PostalCode                                         City   Latitude  \
2        19102  Mid City East,Middle City East,Philadelphia  39.952962   
3        19103  Mid City West,Middle City West,Philadelphia  39.952162   
4        19104                                 Philadelphia  39.961612   
6        19106                                 Philadelphia  39.951062   
7        19107                                 Philadelphia  39.952112   
11       19111                                 Philadelphia  40.057661   
12       19112                                 Philadelphia  39.895677   
13       19114                                 Philadelphia  40.064257   
14       19115                                 Philadelphia  40.092610   
15       19116                                 Philadelphia  40.117413   
16       19118                                 Philadelphia  40.072360   
17       19119                                 Philadelphia  40.053511   
18       19120                                 Philadelphia  40.033944   
19       19121                                 Philadelphia  39.981062   
20       19122                                 Philadelphia  39.977662   
21       19123                                 Philadelphia  39.964012   
22       19124                                 Philadelphia  40.017362   
23       19125                                 Philadelphia  39.978162   
24       19126                                 Philadelphia  40.055411   
25       19127                        Manayunk,Philadelphia  40.026626   
26       19128                                 Philadelphia  40.038944   
27       19129                                 Philadelphia  40.011562   
28       19130                                 Philadelphia  39.968262   
29       19131                                 Philadelphia  39.981112   
30       19132                                 Philadelphia  39.995412   
31       19133                                 Philadelphia  39.992862   
32       19134                                 Philadelphia  39.991712   
33       19135                                 Philadelphia  40.023611   
34       19136                                 Philadelphia  40.041111   
35       19137                                 Philadelphia  40.000262   
36       19138                                 Philadelphia  40.055861   
37       19139                                 Philadelphia  39.961812   
38       19140                                 Philadelphia  40.012212   
39       19141                                 Philadelphia  40.035778   
40       19142                                 Philadelphia  39.922612   
41       19143                                 Philadelphia  39.944162   
42       19144                                 Philadelphia  40.034111   
43       19145                                 Philadelphia  39.922262   
44       19146                                 Philadelphia  39.938512   
45       19147                                 Philadelphia  39.936562   
46       19148                                 Philadelphia  39.919812   
47       19149                                 Philadelphia  40.037711   
48       19150                                 Philadelphia  40.072260   
49       19151                 Overbrook Hills,Philadelphia  39.975929   
50       19152                                 Philadelphia  40.059611   
51       19153                                 Philadelphia  39.898985   
52       19154                                 Philadelphia  40.091460   

    Longitude    Population  
2   -75.16558  Philadelphia  
3   -75.17406  Philadelphia  
4   -75.19957  Philadelphia  
6   -75.14589  Philadelphia  
7   -75.15853  Philadelphia  
11  -75.08018  Philadelphia  
12  -75.19044  Philadelphia  
13  -75.00155  Philadelphia  
14  -75.04118  Philadelphia  
15  -75.01540  Philadelphia  
16  -75.20772  Philadelphia  
17  -75.18858  Philadelphia  
18  -75.12118  Philadelphia  
19  -75.17450  Philadelphia  
2

## 7. Submit Foursquare credentials

In [34]:
# @hidden_cell
# define Foursquare Credentials and Version
CLIENT_ID = 'PKVKP4RSK1UDWBSX241ZZ2BFGZ2055XNG13MU1OZ5G4HO0FY' # your Foursquare ID
CLIENT_SECRET = 'IGWQRC4QT0EDQ0SSRIB4MEURVPRJ2TYL2LJP1BK55TZVGDIU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: PKVKP4RSK1UDWBSX241ZZ2BFGZ2055XNG13MU1OZ5G4HO0FY
CLIENT_SECRET:IGWQRC4QT0EDQ0SSRIB4MEURVPRJ2TYL2LJP1BK55TZVGDIU


In [35]:
first_nei = df_merged['PostalCode'][16]
first_nei

19118

## 8.  We look at all Philadelphia Zip codes, using 19118 (Chestnut Hill) as a reference point.

In [36]:
first_nei_lat = df_merged.loc[16,'Latitude']
first_nei_lon = df_merged.loc[16,'Longitude']
print('Latitude and longitude values of {} are {}, {}.'.format(first_nei, 
                                                               first_nei_lat, 
                                                               first_nei_lon))


Latitude and longitude values of 19118 are 40.072359999999996, -75.20772.


In [37]:
radius = 500 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    first_nei_lat, 
    first_nei_lon, 
    radius, 
    LIMIT)

In [38]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d30e3f392e7a9002c9f54da'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Chestnut Hill',
  'headerFullLocation': 'Chestnut Hill, Philadelphia',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 47,
  'suggestedBounds': {'ne': {'lat': 40.0768600045, 'lng': -75.20185040460325},
   'sw': {'lat': 40.067859995499994, 'lng': -75.21358959539674}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b8ee0e0f964a5204a3c33e3',
       'name': 'Pastorius Park',
       'location': {'address': 'Lincoln Dr',
        'lat': 40.070748284050914,
        'lng': -75.20506381988525,
        'labeledLatLngs': [{'

In [39]:
results['response']['groups'][0]['items'][0]['venue']['categories'][0]['name']

'Park'

In [40]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['reasons.count', 'reasons.items', 'referralId', 'venue.categories',
       'venue.delivery.id', 'venue.delivery.provider.icon.name',
       'venue.delivery.provider.icon.prefix',
       'venue.delivery.provider.icon.sizes', 'venue.delivery.provider.name',
       'venue.delivery.url', 'venue.id', 'venue.location.address',
       'venue.location.cc', 'venue.location.city', 'venue.location.country',
       'venue.location.crossStreet', 'venue.location.distance',
       'venue.location.formattedAddress', 'venue.location.labeledLatLngs',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.neighborhood', 'venue.location.postalCode',
       'venue.location.state', 'venue.name', 'venue.photos.count',
       'venue.photos.groups', 'venue.venuePage.id'],
      dtype='object')

## 9.  Next, we gather information about the venues in Zip Code 19118 from Foursquare.

In [41]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [42]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]
nearby_venues

venue.name  \
0                                      Pastorius Park   
1                     Weavers Way Co-Op Chestnut Hill   
2                      Iron Hill Brewery & Restaurant   
3                                    The Fresh Market   
4                                          El Poquito   
5                                    Campbell's Place   
6                                 Bredenbeck's Bakery   
7                                                Cake   
8                              Bredenbeck's Ice Cream   
9                                 Chestnut Hill Grill   
10                           Paris Bistro & Jazz Cafe   
11                      Chestnut Hill Brewing Company   
12                                          Starbucks   
13                   The Chestnut Hill Farmers Market   
14                                        Flying Fish   
15                          Chestnut Hill Cheese Shop   
16              Ten Thousand Villages - Chestnut Hill   
17                                              Osaka   
18                                  A Taste of Philly   
19                                     Hideaway Music   
20                                      Clover Market   
21                                  Chill On The Hill   
22                                   Fiesta Pizza III   
23                       Robertson's Flowers & Events   
24                                            TD Bank   
25                                               Mica   
26                        Jos. A. Bank Clothiers Inc.   
27                                      King's Garden   
28                          Greene Street Consignment   
29                                        Wells Fargo   
30                                 3000BC WellMed Spa   
31                                 Roller's Express-O   
32                                  The Bone Appetite   
33                                 NEST Chestnut Hill   
34                              Weavers Way Next Door   
35                              Chestnut Hill Bootery   
36                             Artisans on the Avenue   
37                                       J.McLaughlin   
38                                      Stagecrafters   
39                   SEPTA Chestnut Hill West Station   
40  Phantastic Phinds Antique Vintage & Used Furni...   
41                                 Oxford Circus Toys   
42                     Chestnut Hill West End Of Line   
43                            Balance | Chestnut Hill   
44                                  Saigon Noodle Bar   
45                                            Calypso   
46                              Market at the Fareway   

                                     venue.categories  venue.location.lat  \
0   [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           40.070748   
1   [{'id': '4bf58dd8d48988d118951735', 'name': 'G...           40.075489   
2   [{'id': '50327c8591d4c4b30a586d5d', 'name': 'B...           40.075210   
3   [{'id': '4bf58dd8d48988d118951735', 'name': 'G...           40.073300   
4   [{'id': '4bf58dd8d48988d1c1941735', 'name': 'M...           40.073596   
5   [{'id': '4bf58dd8d48988d14e941735', 'name': 'A...           40.075084   
6   [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...           40.072672   
7   [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...           40.076269   
8   [{'id': '4bf58dd8d48988d1c9941735', 'name': 'I...           40.072744   
9   [{'id': '4bf58dd8d48988d14e941735', 'name': 'A...           40.074060   
10  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...           40.074218   
11  [{'id': '50327c8591d4c4b30a586d5d', 'name': 'B...           40.074255   
12  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           40.076419   
13  [{'id': '4bf58dd8d48988d1fa941735', 'name': 'F...           40.074216   
14  [{'id': '4bf58dd8d48988d1ce941735', 'name': 'S...           40.073209   
15  [{'id': '4bf58dd8d48988d11e951735', 'name': 'C...           40.076193 

In [43]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name              categories  \
0                                      Pastorius Park                    Park   
1                     Weavers Way Co-Op Chestnut Hill           Grocery Store   
2                      Iron Hill Brewery & Restaurant                 Brewery   
3                                    The Fresh Market           Grocery Store   
4                                          El Poquito      Mexican Restaurant   
5                                    Campbell's Place     American Restaurant   
6                                 Bredenbeck's Bakery                  Bakery   
7                                                Cake                  Bakery   
8                              Bredenbeck's Ice Cream          Ice Cream Shop   
9                                 Chestnut Hill Grill     American Restaurant   
10                           Paris Bistro & Jazz Cafe       French Restaurant   
11                      Chestnut Hill Brewing Company                 Brewery   
12                                          Starbucks             Coffee Shop   
13                   The Chestnut Hill Farmers Market          Farmers Market   
14                                        Flying Fish      Seafood Restaurant   
15                          Chestnut Hill Cheese Shop             Cheese Shop   
16              Ten Thousand Villages - Chestnut Hill     Arts & Crafts Store   
17                                              Osaka        Sushi Restaurant   
18                                  A Taste of Philly             Snack Place   
19                                     Hideaway Music             Record Shop   
20                                      Clover Market                  Market   
21                                  Chill On The Hill          Ice Cream Shop   
22                                   Fiesta Pizza III             Pizza Place   
23                       Robertson's Flowers & Events             Flower Shop   
24                                            TD Bank                    Bank   
25                                               Mica     American Restaurant   
26                        Jos. A. Bank Clothiers Inc.             Men's Store   
27                                      King's Garden      Chinese Restaurant   
28                          Greene Street Consignment          Clothing Store   
29                                        Wells Fargo                    Bank   
30                                 3000BC WellMed Spa                     Spa   
31                                 Roller's Express-O                    Café   
32                                  The Bone Appetite               Pet Store   
33                                 NEST Chestnut Hill      Child Care Service   
34                              Weavers Way Next Door          Cosmetics Shop   
35                              Chestnut Hill Bootery              Shoe Store   
36                             Artisans on the Avenue                Boutique   
37                                       J.McLaughlin          Clothing Store   
38                                      Stagecrafters                 Theater   
39                   SEPTA Chestnut Hill West Station           Train Station   
40  Phantastic Phinds Antique Vintage & Used Furni...  Furniture / Home Store   
41                                 Oxford Circus Toys        Toy / Game Store   
42                     Chestnut Hill West End Of Line           Train Station   
43                            Balance | Chestnut Hill    Gym / Fitness Center   
44                                  Saigon Noodle Bar            Noodle House   
45                                            Calypso    Caribbean Restaurant   
46                              Market at the Fareway                  Market   

          lat        lng  
0   40.070748 -75.205064  
1   40.075489 -75.205659  
2   40.075210 -75.205261  
3   40.073300 -75.203366  
4   40.073596 -75.202887 

## 10. We retrieve lists of venues in each Philadelphia Zip code.

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [45]:
phil_venues = getNearbyVenues(names = df_merged['PostalCode'],
                                   latitudes = df_merged['Latitude'],
                                   longitudes = df_merged['Longitude']
                                  )

19102
19103
19104
19106
19107
19111
19112
19114
19115
19116
19118
19119
19120
19121
19122
19123
19124
19125
19126
19127
19128
19129
19130
19131
19132
19133
19134
19135
19136
19137
19138
19139
19140
19141
19142
19143
19144
19145
19146
19147
19148
19149
19150
19151
19152
19153
19154


In [47]:
print(phil_venues.shape)
phil_venues.head(47)

(1158, 7)


PostalCode  Neighborhood Latitude  Neighborhood Longitude  \
0        19102              39.952962               -75.16558   
1        19102              39.952962               -75.16558   
2        19102              39.952962               -75.16558   
3        19102              39.952962               -75.16558   
4        19102              39.952962               -75.16558   
5        19102              39.952962               -75.16558   
6        19102              39.952962               -75.16558   
7        19102              39.952962               -75.16558   
8        19102              39.952962               -75.16558   
9        19102              39.952962               -75.16558   
10       19102              39.952962               -75.16558   
11       19102              39.952962               -75.16558   
12       19102              39.952962               -75.16558   
13       19102              39.952962               -75.16558   
14       19102              39.952962               -75.16558   
15       19102              39.952962               -75.16558   
16       19102              39.952962               -75.16558   
17       19102              39.952962               -75.16558   
18       19102              39.952962               -75.16558   
19       19102              39.952962               -75.16558   
20       19102              39.952962               -75.16558   
21       19102              39.952962               -75.16558   
22       19102              39.952962               -75.16558   
23       19102              39.952962               -75.16558   
24       19102              39.952962               -75.16558   
25       19102              39.952962               -75.16558   
26       19102              39.952962               -75.16558   
27       19102              39.952962               -75.16558   
28       19102              39.952962               -75.16558   
29       19102              39.952962               -75.16558   
30       19102              39.952962               -75.16558   
31       19102              39.952962               -75.16558   
32       19102              39.952962               -75.16558   
33       19102              39.952962               -75.16558   
34       19102              39.952962               -75.16558   
35       19102              39.952962               -75.16558   
36       19102              39.952962               -75.16558   
37       19102              39.952962               -75.16558   
38       19102              39.952962               -75.16558   
39       19102              39.952962               -75.16558   
40       19102              39.952962               -75.16558   
41       19102              39.952962               -75.16558   
42       19102              39.952962               -75.16558   
43       19102              39.952962               -75.16558   
44       19102              39.952962               -75.16558   
45       19102              39.952962               -75.16558   
46       19102              39.952962               -75.16558   

                                          Venue  Venue Latitude  \
0                                 Dilworth Park       39.952772   
1                    La Colombe Coffee Roasters       39.951659   
2                           City Hall Courtyard       39.952484   
3                  One Liberty Observation Deck       39.952740   
4                         JFK Plaza / Love Park       39.954123   
5                The Ritz-Carlton, Philadelphia       39.951446   
6                                    sweetgreen       39.953115   
7                                        UNIQLO       39.951426   
8         Del Frisco's Double Eagle Steak House       39.950795   
9                                           R2L       39.951714   
10                     Philadelphia Film Center       39.950835   
11  Rothman Institute Ice Rink at Dilworth Park       39.952770   
12          

In [48]:
phil_venues.groupby('PostalCode').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
PostalCode                                                         
19102                         100                     100    100   
19103                         100                     100    100   
19104                          15                      15     15   
19106                         100                     100    100   
19107                         100                     100    100   
19111                          14                      14     14   
19112                           1                       1      1   
19114                           6                       6      6   
19115                           3                       3      3   
19116                          13                      13     13   
19118                          47                      47     47   
19119                          21                      21     21   
19120                          22                      22     22   
19121                          15                      15     15   
19122                          16                      16     16   
19123                          37                      37     37   
19124                           8                       8      8   
19125                          25                      25     25   
19126                           9                       9      9   
19127                          61                      61     61   
19128                          17                      17     17   
19129                          11                      11     11   
19130                          24                      24     24   
19131                          31                      31     31   
19132                           9                       9      9   
19133                          16                      16     16   
19134                          11                      11     11   
19135                           9                       9      9   
19136                          18                      18     18   
19137                          15                      15     15   
19138                           4                       4      4   
19139                           9                       9      9   
19140                           8                       8      8   
19141                          11                      11     11   
19142                           5                       5      5   
19143                           6                       6      6   
19144                          22                      22     22   
19145                          40                      40     40   
19146                          22                      22     22   
19147                          75                      75     75   
19148                          13                      13     13   
19149                          18                      18     18   
19150                           4                       4      4   
19151                          11                      11     11   
19152                          26                      26     26   
19153                           4                       4      4   
19154                           6                       6      6   

            Venue Latitude  Venue Longitude  Venue Category  
PostalCode                                                   
19102                  100              100             100  
19103                  100              100             100  
19104                   15               15              15  
19106                  100              100             100  
19107                  100              100             100  
19111                   14               14              14  
19112                    1                1               1  
19114                    6                6               6  
19115                    3                3               3

In [49]:
print('There are {} unique categories.'.format(len(phil_venues['Venue Category'].unique())))

There are 221 unique categories.


In [50]:
phil_onehot = pd.get_dummies(phil_venues[['Venue Category']], prefix="", prefix_sep="")

phil_onehot['PostalCode'] = phil_venues['PostalCode'] 

fixed_columns = [phil_onehot.columns[-1]] + list(phil_onehot.columns[:-1])
phil_onehot = phil_onehot[fixed_columns]
phil_onehot.head(47)

PostalCode  African Restaurant  American Restaurant  Art Gallery  \
0        19102                   0                    0            0   
1        19102                   0                    0            0   
2        19102                   0                    0            0   
3        19102                   0                    0            0   
4        19102                   0                    0            0   
5        19102                   0                    0            0   
6        19102                   0                    0            0   
7        19102                   0                    0            0   
8        19102                   0                    0            0   
9        19102                   0                    1            0   
10       19102                   0                    0            0   
11       19102                   0                    0            0   
12       19102                   0                    0            0   
13       19102                   0                    0            0   
14       19102                   0                    0            0   
15       19102                   0                    0            0   
16       19102                   0                    0            0   
17       19102                   0                    0            0   
18       19102                   0                    0            0   
19       19102                   0                    0            0   
20       19102                   0                    0            0   
21       19102                   0                    0            0   
22       19102                   0                    0            0   
23       19102                   0                    0            0   
24       19102                   0                    0            0   
25       19102                   0                    0            0   
26       19102                   0                    1            0   
27       19102                   0                    0            0   
28       19102                   0                    0            0   
29       19102                   0                    0            0   
30       19102                   0                    0            0   
31       19102                   0                    0            0   
32       19102                   0                    0            0   
33       19102                   0                    0            0   
34       19102                   0                    0            0   
35       19102                   0                    0            0   
36       19102                   0                    0            0   
37       19102                   0                    0            0   
38       19102                   0                    0            0   
39       19102                   0                    0            0   
40       19102                   0                    0            0   
41       19102                   0                    0            0   
42       19102                   0                    0            0   
43       19102                   0                    0            0   
44       19102                   0                    0            0   
45       19102                   0                    0            0   
46       19102                   0                    0            0   

    Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0            0                    0                     0                 0   
1            0                    0                     0                 0   
2            0                    0                     0                 0   
3            0                    0                     0                 0   
4            0                    0                     0                 0   
5            0                    0                     0                

In [51]:
phil_grouped = phil_onehot.groupby('PostalCode').mean().reset_index()
phil_grouped

PostalCode  African Restaurant  American Restaurant  Art Gallery  \
0        19102                 0.0             0.040000     0.000000   
1        19103                 0.0             0.060000     0.000000   
2        19104                 0.0             0.000000     0.066667   
3        19106                 0.0             0.040000     0.020000   
4        19107                 0.0             0.010000     0.010000   
5        19111                 0.0             0.000000     0.000000   
6        19112                 0.0             0.000000     0.000000   
7        19114                 0.0             0.000000     0.000000   
8        19115                 0.0             0.000000     0.000000   
9        19116                 0.0             0.000000     0.000000   
10       19118                 0.0             0.063830     0.000000   
11       19119                 0.0             0.047619     0.000000   
12       19120                 0.0             0.000000     0.000000   
13       19121                 0.0             0.000000     0.266667   
14       19122                 0.0             0.062500     0.000000   
15       19123                 0.0             0.027027     0.000000   
16       19124                 0.0             0.000000     0.000000   
17       19125                 0.0             0.000000     0.080000   
18       19126                 0.0             0.000000     0.000000   
19       19127                 0.0             0.016393     0.000000   
20       19128                 0.0             0.000000     0.000000   
21       19129                 0.0             0.000000     0.000000   
22       19130                 0.0             0.125000     0.000000   
23       19131                 0.0             0.000000     0.000000   
24       19132                 0.0             0.000000     0.000000   
25       19133                 0.0             0.000000     0.062500   
26       19134                 0.0             0.090909     0.090909   
27       19135                 0.0             0.000000     0.000000   
28       19136                 0.0             0.000000     0.000000   
29       19137                 0.0             0.066667     0.000000   
30       19138                 0.0             0.000000     0.000000   
31       19139                 0.0             0.000000     0.000000   
32       19140                 0.0             0.000000     0.000000   
33       19141                 0.0             0.000000     0.000000   
34       19142                 0.2             0.000000     0.000000   
35       19143                 0.0             0.000000     0.000000   
36       19144                 0.0             0.090909     0.000000   
37       19145                 0.0             0.000000     0.000000   
38       19146                 0.0             0.000000     0.000000   
39       19147                 0.0             0.013333     0.026667   
40       19148                 0.0             0.000000     0.000000   
41       19149                 0.0             0.055556     0.000000   
42       19150                 0.0             0.000000     0.000000   
43       19151                 0.0             0.090909     0.000000   
44       19152                 0.0             0.000000     0.000000   
45       19153                 0.0             0.000000     0.000000   
46       19154                 0.0             0.000000     0.000000   

    Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0         0.01             0.010000                0.0000          0.000000   
1         0.00             0.000000                0.0000          0.010000   
2         0.00             0.000000                0.0000          0.000000   
3         0.00             0.000000                0.0000          0.010000   
4         0.00             0.000000                0.0000          0.010000   
5         0.00             0.000000                0.0000          0.0000

In [52]:
phil_grouped.PostalCode = phil_grouped.PostalCode.astype(str)

## 11. Identify the most common venues for each Zip code.

In [53]:
num_top_venues = 5
for code in phil_grouped['PostalCode']:
    print("----"+code+"----")
    temp = phil_grouped[phil_grouped['PostalCode'] == code].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----19102----
                 venue  freq
0          Coffee Shop  0.07
1                Hotel  0.06
2  American Restaurant  0.04
3          Salad Place  0.04
4   Italian Restaurant  0.03


----19103----
                           venue  freq
0            American Restaurant  0.06
1                    Coffee Shop  0.05
2                  Deli / Bodega  0.04
3  Vegetarian / Vegan Restaurant  0.03
4        New American Restaurant  0.03


----19104----
                  venue  freq
0      Toy / Game Store  0.07
1             Piano Bar  0.07
2      Greek Restaurant  0.07
3  Caribbean Restaurant  0.07
4     Mobile Phone Shop  0.07


----19106----
                 venue  freq
0       History Museum  0.08
1        Historic Site  0.06
2          Coffee Shop  0.05
3  American Restaurant  0.04
4             Boutique  0.04


----19107----
                venue  freq
0              Bakery  0.06
1      Sandwich Place  0.05
2        Burger Joint  0.03
3  Chinese Restaurant  0.03
4               Hote

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [55]:
phil_grouped.PostalCode = phil_grouped.PostalCode.astype(int)

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = phil_grouped['PostalCode']

for ind in np.arange(phil_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(phil_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0        19102           Coffee Shop                 Hotel   
1        19103   American Restaurant           Coffee Shop   
2        19104         Deli / Bodega  Caribbean Restaurant   
3        19106        History Museum         Historic Site   
4        19107                Bakery        Sandwich Place   
5        19111         Deli / Bodega        Baseball Field   
6        19112            Food Truck           Event Space   
7        19114           Golf Course   Sporting Goods Shop   
8        19115                  Pool    Italian Restaurant   
9        19116    Chinese Restaurant           Beer Garden   
10       19118   American Restaurant         Grocery Store   
11       19119           Pizza Place              Pharmacy   
12       19120    Seafood Restaurant            Donut Shop   
13       19121           Art Gallery         Deli / Bodega   
14       19122               Brewery                  Park   
15       19123           Coffee Shop            Donut Shop   
16       19124   Sporting Goods Shop     Convenience Store   
17       19125                   Bar           Pizza Place   
18       19126     Convenience Store            Donut Shop   
19       19127                   Bar         Grocery Store   
20       19128                 Track       Thai Restaurant   
21       19129  Gym / Fitness Center                   Pub   
22       19130   American Restaurant                   Bar   
23       19131        Clothing Store                  Park   
24       19132         Grocery Store      Video Game Store   
25       19133  Fast Food Restaurant   Fried Chicken Joint   
26       19134        Discount Store            Donut Shop   
27       19135                   Pub              Pharmacy   
28       19136          Intersection        Discount Store   
29       19137           Pizza Place              Pharmacy   
30       19138        Ice Cream Shop            Restaurant   
31       19139                Lounge                  Food   
32       19140  Fast Food Restaurant           Pizza Place   
33       19141            Food Truck         Metro Station   
34       19142    African Restaurant              Bus Stop   
35       19143           Supermarket          Intersection   
36       19144   American Restaurant   Fried Chicken Joint   
37       19145           Pizza Place        Discount Store   
38       19146    Mexican Restaurant        Discount Store   
39       19147    Mexican Restaurant    Italian Restaurant   
40       19148    Mexican Restaurant          Liquor Store   
41       19149     Convenience Store             Nightclub   
42       19150         Deli / Bodega    Athletics & Sports   
43       19151         Deli / Bodega        Cosmetics Shop   
44       19152           Pizza Place              Pharmacy   
45       19153                 Hotel   Rental Car Location   
46       19154           Pizza Place         Deli / Bodega   

              3rd Most Common Venue 4th Most Common Venue  \
0                       Salad Place   American Restaurant   
1                     Deli / Bodega                   Bar   
2                       Coffee Shop    Chinese Restaurant   
3                       Coffee Shop   American Restaurant   
4                Chinese Restaurant                 Hotel   
5                 Convenience Store           Bus Station   
6                        Food Court     Food & Drink Shop   
7                    Discount Store            Donut Shop   
8        Construction & Landscaping  Ethiopian Restaurant   
9                              Bank           Flower Shop   
10                   Ice Cream Shop        Clothing Store   
11               Chinese Restaurant         Train Station   
12                       Kids Store                   Bar   
13                       Restaurant          Intersection   
14             Colombian Restaurant         Grocery Store   
15                            Diner              Ph

## 12.  Identify clusters within the data.

In [56]:
from sklearn.cluster import KMeans

kclusters = 5

phil_grouped_clustering = phil_grouped.drop('PostalCode', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(phil_grouped_clustering)

kmeans.labels_

array([2, 2, 2, 2, 2, 2, 1, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 0,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2,
       2, 2, 3], dtype=int32)

In [57]:
phil_merged = df_merged[0:47]

phil_merged['Cluster Labels'] = kmeans.labels_

phil_merged = phil_merged.join(neighborhoods_venues_sorted.set_index('PostalCode'), on='PostalCode')

phil_merged.head(47)

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


PostalCode                                         City   Latitude  \
2        19102  Mid City East,Middle City East,Philadelphia  39.952962   
3        19103  Mid City West,Middle City West,Philadelphia  39.952162   
4        19104                                 Philadelphia  39.961612   
6        19106                                 Philadelphia  39.951062   
7        19107                                 Philadelphia  39.952112   
11       19111                                 Philadelphia  40.057661   
12       19112                                 Philadelphia  39.895677   
13       19114                                 Philadelphia  40.064257   
14       19115                                 Philadelphia  40.092610   
15       19116                                 Philadelphia  40.117413   
16       19118                                 Philadelphia  40.072360   
17       19119                                 Philadelphia  40.053511   
18       19120                                 Philadelphia  40.033944   
19       19121                                 Philadelphia  39.981062   
20       19122                                 Philadelphia  39.977662   
21       19123                                 Philadelphia  39.964012   
22       19124                                 Philadelphia  40.017362   
23       19125                                 Philadelphia  39.978162   
24       19126                                 Philadelphia  40.055411   
25       19127                        Manayunk,Philadelphia  40.026626   
26       19128                                 Philadelphia  40.038944   
27       19129                                 Philadelphia  40.011562   
28       19130                                 Philadelphia  39.968262   
29       19131                                 Philadelphia  39.981112   
30       19132                                 Philadelphia  39.995412   
31       19133                                 Philadelphia  39.992862   
32       19134                                 Philadelphia  39.991712   
33       19135                                 Philadelphia  40.023611   
34       19136                                 Philadelphia  40.041111   
35       19137                                 Philadelphia  40.000262   
36       19138                                 Philadelphia  40.055861   
37       19139                                 Philadelphia  39.961812   
38       19140                                 Philadelphia  40.012212   
39       19141                                 Philadelphia  40.035778   
40       19142                                 Philadelphia  39.922612   
41       19143                                 Philadelphia  39.944162   
42       19144                                 Philadelphia  40.034111   
43       19145                                 Philadelphia  39.922262   
44       19146                                 Philadelphia  39.938512   
45       19147                                 Philadelphia  39.936562   
46       19148                                 Philadelphia  39.919812   
47       19149                                 Philadelphia  40.037711   
48       19150                                 Philadelphia  40.072260   
49       19151                 Overbrook Hills,Philadelphia  39.975929   
50       19152                                 Philadelphia  40.059611   
51       19153                                 Philadelphia  39.898985   
52       19154                                 Philadelphia  40.091460   

    Longitude    Population  Cluster Labels 1st Most Common Venue  \
2   -75.16558  Philadelphia               2           Coffee Shop   
3   -75.17406  Philadelphia               2   American Restaurant   
4   -75.19957  Philadelphia               2         Deli / Bodega   
6   -75.14589  Philadelphia               2        History Museum   
7   -75.15853  Philadelphia               2                Bakery   
11  -75.08018  Philadelphia          

## 13. Generate map with clusters superimposed. 

In [58]:
map_clusters = folium.Map(location=[39.9524152, -75.1635755], zoom_start=11)

x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)

markers_colors = []
for lat, lon, nei , cluster in zip(phil_merged['Latitude'], phil_merged['Longitude'], phil_merged['PostalCode'], phil_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']


You can see the map by clicking on [this link](https://github.com/saracw/Coursera---IBM-Capstone-1/blob/master/PhillyClusters2.png).

In [59]:
phil_merged.loc[phil_merged['Cluster Labels'] == 0, phil_merged.columns[[0] + list(range(5, phil_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
23       19125               0                   Bar           Pizza Place   
27       19129               0  Gym / Fitness Center                   Pub   
46       19148               0    Mexican Restaurant          Liquor Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
23           Art Gallery                  Café              Dive Bar   
27        Massage Studio                  Café     Food & Drink Shop   
46           Pizza Place                  Park    Chinese Restaurant   

   6th Most Common Venue            7th Most Common Venue  \
23           Beer Garden  Southern / Soul Food Restaurant   
27    Rock Climbing Spot                     Dessert Shop   
46    Seafood Restaurant                    Grocery Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23                Market        Sandwich Place           Liquor Store  
27           Fabric Shop                  Food            Flower Shop  
46                   Bar                Bakery   Cambodian Restaurant

In [60]:
phil_merged.loc[phil_merged['Cluster Labels'] == 1, phil_merged.columns[[0] + list(range(5, phil_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12       19112               1            Food Truck           Event Space   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12            Food Court     Food & Drink Shop                  Food   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12           Flower Shop           Fish Market   Filipino Restaurant   

   9th Most Common Venue 10th Most Common Venue  
12                 Field   Fast Food Restaurant

In [61]:
phil_merged.loc[phil_merged['Cluster Labels'] == 2, phil_merged.columns[[0] + list(range(5, phil_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2        19102               2           Coffee Shop                 Hotel   
3        19103               2   American Restaurant           Coffee Shop   
4        19104               2         Deli / Bodega  Caribbean Restaurant   
6        19106               2        History Museum         Historic Site   
7        19107               2                Bakery        Sandwich Place   
11       19111               2         Deli / Bodega        Baseball Field   
13       19114               2           Golf Course   Sporting Goods Shop   
15       19116               2    Chinese Restaurant           Beer Garden   
16       19118               2   American Restaurant         Grocery Store   
17       19119               2           Pizza Place              Pharmacy   
18       19120               2    Seafood Restaurant            Donut Shop   
19       19121               2           Art Gallery         Deli / Bodega   
20       19122               2               Brewery                  Park   
21       19123               2           Coffee Shop            Donut Shop   
22       19124               2   Sporting Goods Shop     Convenience Store   
24       19126               2     Convenience Store            Donut Shop   
25       19127               2                   Bar         Grocery Store   
26       19128               2                 Track       Thai Restaurant   
28       19130               2   American Restaurant                   Bar   
29       19131               2        Clothing Store                  Park   
30       19132               2         Grocery Store      Video Game Store   
31       19133               2  Fast Food Restaurant   Fried Chicken Joint   
32       19134               2        Discount Store            Donut Shop   
33       19135               2                   Pub              Pharmacy   
34       19136               2          Intersection        Discount Store   
35       19137               2           Pizza Place              Pharmacy   
36       19138               2        Ice Cream Shop            Restaurant   
37       19139               2                Lounge                  Food   
38       19140               2  Fast Food Restaurant           Pizza Place   
39       19141               2            Food Truck         Metro Station   
40       19142               2    African Restaurant              Bus Stop   
41       19143               2           Supermarket          Intersection   
42       19144               2   American Restaurant   Fried Chicken Joint   
43       19145               2           Pizza Place        Discount Store   
44       19146               2    Mexican Restaurant        Discount Store   
45       19147               2    Mexican Restaurant    Italian Restaurant   
47       19149               2     Convenience Store             Nightclub   
48       19150               2         Deli / Bodega    Athletics & Sports   
49       19151               2         Deli / Bodega        Cosmetics Shop   
50       19152               2           Pizza Place              Pharmacy   
51       19153               2                 Hotel   Rental Car Location   

              3rd Most Common Venue 4th Most Common Venue  \
2                       Salad Place   American Restaurant   
3                     Deli / Bodega                   Bar   
4                       Coffee Shop    Chinese Restaurant   
6                       Coffee Shop   American Restaurant   
7                Chinese Restaurant                 Hotel   
11                Convenience Store           Bus Station   
13                   Discount Store            Donut Shop   
15                             Bank           Flower Shop   
16                   Ice Cream Shop        Clothing Store   
17               Chinese Restaurant         Train Station   
18                       Kids Store                   Ba

In [62]:
phil_merged.loc[phil_merged['Cluster Labels'] == 3, phil_merged.columns[[0] + list(range(5, phil_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
52       19154               3           Pizza Place         Deli / Bodega   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
52     Convenience Store              Pharmacy           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
52     Food & Drink Shop                  Food           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
52           Fish Market    Filipino Restaurant

In [63]:
phil_merged.loc[phil_merged['Cluster Labels'] == 4, phil_merged.columns[[0] + list(range(5, phil_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14       19115               4                  Pool    Italian Restaurant   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14  Construction & Landscaping  Ethiopian Restaurant     Food & Drink Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14                  Food           Flower Shop           Fish Market   

   9th Most Common Venue 10th Most Common Venue  
14   Filipino Restaurant                  Field

## 14. Overview of results

As you can see, above, I was able to cluster Philadelphia's Zip codes using k-means clustering. However, what you might not know if you are unfamiliar with this city, is that there the clustering does not yield great insights. This is, in part, because Zip codes exist for the sole purpose of facilitating delivery of the mail. Below, you will see a Zip code map of Philadelphia, with the clusters color coded to indicate whether they fall into cluster(s) 0, 1, 2, 3, or 4. 

[Color-coded map of Philadelphia clusters](https://github.com/saracw/Coursera---IBM-Capstone-1/blob/master/CHillvenues.csv)

## 15. Discussion

>The primary purpose of this exercise was to determine whether we were able to use what we learned during the course of this Specialization, independently and without any Lab to provide explicit instructions. In that, this project was successful. I was able to run code that produced a coherent result. I also got to learn more about my own Philadelphia neighborhood, since that was the topic upon which I chose to focus. 

>Still, I set out to answer another question, which was "If I were to move to Toronto, is there a neighborhood that is similar to mine, in a location that would also be within walking distance to and area that could support a growing Consulting/Data Science business? The characteristics of such a location would include mixed residential and commercial use, in a neighborhood that is roughly as old as my own (founded before 1800), with trees and good places to go (again, on foot) as well as large business's headquarters and growing tech investment.

>In both cases, I used Foursquare data to assess the number and type of venues in both cities. However, Foursquare was developed to help people pick somewhere to go, based on their wishes and their location. It was not designed to help people make momentous life decisions. That said, venues matter - and sometimes they can tell you something about a neighborhood that you might like or really not like. I found this to be the case when I first collected a list of all the venues in my neighborhood (Philadelphia's Chestnut Hill, Zip code 19118) and the Foursquare Toronto data, parsed by Zip code, latitude and longitude. Please see the spreadsheets I downloaded from my Jupyter notebook, below. 

In [64]:
import types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share your notebook.
client_956b1639be4e4261b6195120965e82e4 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='KxwGJCkXCxDqMYElNdYMwFO1iAOhP2IyaLqjubvpZg3i',
    ibm_auth_endpoint="https://iam.ng.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

body = client_956b1639be4e4261b6195120965e82e4.get_object(Bucket='neighborhoodsampcitiesphiladelphi-donotdelete-pr-46wtikf2skx3m1',Key='CHillvenues.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head(46)

Venue Number                              Venue Name  \
0              5                        Campbell's Place   
1              9                     Chestnut Hill Grill   
2             25                                    Mica   
3             16   Ten Thousand Villages - Chestnut Hill   
4              6                     Bredenbeck's Bakery   
5              7                                    Cake   
6             24                                 TD Bank   
7             29                             Wells Fargo   
8             34                  Artisans on the Avenue   
9              2          Iron Hill Brewery & Restaurant   
10            11           Chestnut Hill Brewing Company   
11            31                      Roller's Express-O   
12            45                                 Calypso   
13            15               Chestnut Hill Cheese Shop   
14            27                           King's Garden   
15            28               Greene Street Consignment   
16            35                            J.McLaughlin   
17            12                               Starbucks   
18            40                            Poppy's Cafe   
19            33                   Weavers Way Next Door   
20            38        Drake's Gourmet Foods & Catering   
21            13        The Chestnut Hill Farmers Market   
22            23            Robertson's Flowers & Events   
23            10                Paris Bistro & Jazz Cafe   
24             1         Weavers Way Co-Op Chestnut Hill   
25             3                        The Fresh Market   
26             8                  Bredenbeck's Ice Cream   
27            21                       Chill On The Hill   
28            37                            The Knit Wit   
29            20                           Clover Market   
30            26             Jos. A. Bank Clothiers Inc.   
31             4                              El Poquito   
32            41                             The Paperia   
33             0                          Pastorius Park   
34            32                       The Bone Appetite   
35            22                        Fiesta Pizza III   
36            19                          Hideaway Music   
37            14                             Flying Fish   
38            18                       A Taste of Philly   
39            30                      3000BC WellMed Spa   
40            17                                   Osaka   
41            44  Sue's Custom Dressmaking & Tailor Shop   
42            36                           Stagecrafters   
43            43                      Oxford Circus Toys   
44            39        SEPTA Chestnut Hill West Station   

                   Venue Category  Zip Code   Neighborhood   Latitude  \
0             American Restaurant     19118  Chestnut Hill  40.075084   
1             American Restaurant     19118  Chestnut Hill  40.074060   
2    American - Fusion Restaurant     19118  Chestnut Hill  40.076777   
3             Arts & Crafts Store     19118  Chestnut Hill  40.075044   
4                          Bakery     19118  Chestnut Hill  40.072672   
5                          Bakery     19118  Chestnut Hill  40.076269   
6                            Bank     19118  Chestnut Hill  40.076596   
7                            Bank     19118  Chestnut Hill  40.076703   
8                        Boutique     19118  Chestnut Hill  40.075755   
9                         Brewery     19118  Chestnut Hill  40.075210   
10                        Brewery     19118  Chestnut Hill  40.074255   
11                           Caf_     19118  Chestnut Hill  40.075125   
12           Caribbean Restaurant     19118  Chestnut Hill  40.074165   
13                    Cheese Shop     19118  Chestnut Hill  40.076193   
14             Chinese Restaurant     19118  Chestnut Hill  40.073819   
15                 Clothing Store     19118  Chestnut Hill  40.076462   
16

## Toronto Venues

In [ ]:
# body = client_956b1639be4e4261b6195120965e82e4.get_object(Bucket='neighborhoodsampcitiesphiladelphi-donotdelete-pr-46wtikf2skx3m1',Key='torontovenues.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_2 = pd.read_csv(body)
df_data_2.head(38)


## 16. Implications and Conclusions

>Relocation analysis is serious business, and this data collection/analysis process is a good beginning. Going forward, I plan to use it as a jumping off point for looking at neighborhoods, using units of measurement (e.g., the Census block group that are more stable and are linked with larger data sets like the Economic Census, as well as differences in governance, etc. - which are a function of differences between Canada and the US. One thing I know and like very much is Canada's approach to immigration (in other words, it is good for society, the economy, the wellbeing of all and should be encouraged).

>Still, I did find a partial answer to my question of where I would want to live/work if I moved to Toronto. "Old Toronto" looks very attractive to me for so many reasons - not the least of which is the presence of the University nearby. Thriving educational institutions are essential to a good economy - especially if the type of work one does is cognitive in nature, as is the case for me. This is a pretty large area and includes a broad variety of neighborhoods. 

>One of the analyses I conducted was to look at the mix of venues in my own neighborhood, then sort the Toronto data to see which among the neigborhods covered in our class's work was most similar to my own. One thing I noticed: Chestnut Hill likes food! There are several ice cream parlors and bakeries - moreseo than in many, larger neighborhoods. There is also a farmer's market, and two of the three grocery grocery stores are organic. There are numerous parkes, one of which is among the largest in the US, and there are two light commuter rail lines. Among the Toronto Postal Codes we looked at, the Toronto neighbourhood called "Harbord," near the University of Toronto, seems to have all of these things. Yes, more research is required, but now I have more and better tools at my disposal. 

## Thank you very much for your time and interest!